src/SUnit-Tests/ProcessMonitorTestServiceTest.class.st

```diff
@@ -0,0 +1,556 @@
+	| executed |
+	executed := false.
+	self fork: [ testService suspendBackgroundFailure: Error new. executed := true ].
+
+	testService handleException: Halt new. 
+	
+	self assert: executed.
+	self deny: testService shouldSuspendBackgroundFailures
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testResumeFailedProcessesWhenHaltIsSignaledInBackground [
+	| processWithHalt executed |
+	executed := false.
+	self fork: [ testService suspendBackgroundFailure: Error new. executed := true ].
+	processWithHalt := self fork: [ Processor activeProcess suspend. Halt new signal ].
+	
+	processWithHalt on: Halt do: [ :actual | ].
+	processWithHalt resume.
+	
+	self assert: executed.
+	self deny: testService shouldSuspendBackgroundFailures
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testResumeFailedProcessesWhenTestFailureIsUnhandled [
+	| executed |
+	executed := false.
+	self fork: [ testService suspendBackgroundFailure: Error new. executed := true ].
+	self deny: executed.	
+
+	testService handleException: Error new. "UnhandledError always started with original error signal"
+	testService handleException: UnhandledError new.
+	self assert: executed.
+	"Unhandled exception opens debugger and therefore any new errors should not be suspended"
+	self deny: testService shouldSuspendBackgroundFailures 
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testShouldSuspendBackgroundFailuresByDefault [
+
+	self assert: testService shouldSuspendBackgroundFailures 
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testStoreAllForkedProcesses [
+
+	| process |
+	process := self fork: [ ].
+	
+	self assert: (testService forkedProcesses includes: process)
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testSuspendBackgroundError [
+
+	| process error suspendedError |
+	error := Error new messageText: 'test error'.
+	
+	process := self fork: [ error signal ].
+	
+	self assert: process isSuspended.
+	suspendedError := testService suspendedBackgroundFailures at: process.
+	self assert: suspendedError class equals: UnhandledError. 
+	self assert: suspendedError exception identicalTo: error.
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testSuspendBackgroundWarning [
+	
+	| process warning suspendedError |
+	warning := Warning new.
+	process := self fork: [ warning signal ].
+	
+	self assert: process isSuspended.
+	suspendedError := testService suspendedBackgroundFailures at: process.
+	self assert: suspendedError class equals: UnhandledError. 
+	self assert: suspendedError exception identicalTo: warning.
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testSuspendResumeTerminateIssue [
+	"The test discovers the issue with Process>>isTerminated method
+	when process was suspended in the middle and then resumed.
+	During the bootstrap process on CI it fails RunKernelTests stage.
+	But it is working on full image"
```

Comment:
This is separate test for the issue. It is important to keep it in this class. It seems the case appears when tests are running in concrete order.